incubator/hnc/internal/reconcilers/object.go

```diff
@@ -359,18 +363,20 @@ func (r *ObjectReconciler) syncPropagated(ctx context.Context, log logr.Logger,
 	if srcInst == nil {
 		return actionRemove, nil
 	}
-
 	// If an object doesn't exist, assume it's been deleted or not yet created.
 	exists := inst.GetCreationTimestamp() != v1.Time{}
-
-	// If the copy does not exist, or is different from the source, return the write action and the
-	// source instance. Note that DeepEqual could return `true` even if the object doesn't exist if
-	// the source object is trivial (e.g. a completely empty ConfigMap).
-	if !exists ||
-		!reflect.DeepEqual(object.Canonical(inst), object.Canonical(srcInst)) ||
-		inst.GetLabels()[api.LabelInheritedFrom] != srcInst.GetNamespace() {
-		metadata.SetLabel(inst, api.LabelInheritedFrom, srcInst.GetNamespace())
-		return actionWrite, srcInst
+	selector := r.getSelector(ctx, log, srcInst)
+
+	if selector == nil || selector.Matches(labels.Set(r.Forest.Get(inst.GetNamespace()).GetLabels())) {
```

Comment:
I think an early exit might be much more appropriate here (e.g. above line 366). This is very hard to read, with two levels of nested `if` statements (and also this line is much too long, IMO - you can/should introduce some temp variables here).

Also - what happens if `srcInst` _used_ to be propagated here, but now has a new selector that prevents it from being propagated here? This will execute line 382 which says "the object already exists and doesn't need to be updated" - does that seem correct?